io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -244,47 +246,27 @@ schedule !thread@Thread{
         -- There is a possible race between timeout action and the timeout expiration.
         -- We use a lock to solve the race.
-        --
-        -- The lock starts 'NotLocked' and when the timeout fires the lock is
-        -- locked and asynchronously an assassin thread is coming to interrupt
-        -- it. If the lock is locked when the timeout is fired then nothing
-        -- happens.
-        --
-        -- Knowing this, if we reached this point in the code and the lock is
-        -- 'Locked', then it means that this thread still hasn't received the
-        -- 'TimeoutException', so we need to kill the thread that is responsible
-        -- for doing that (the assassin thread, we need to defend ourselves!)
-        -- and run our continuation successfully and peacefully. We will do that
-        -- by uninterruptibly-masking ourselves so we can not receive any
-        -- exception and kill the assassin thread behind its back.
-        -- If the lock is 'NotLocked' then it means we can just acquire it and
-        -- carry on with the success case.
-        locked <- readSTRef isLockedRef
-        case locked of
-          Locked etid -> do
-            let -- Kill the assassin throwing thread then unmask exceptions and
-                -- carry on the continuation
-                thread' =
-                  thread { threadControl =
-                            ThreadControl (ThrowTo (toException ThreadKilled)
-                                                   etid
-                                                   (Return ()))
-                                          (MaskFrame (\_ -> k (Just x)) maskst ctl')
-                         , threadMasking = MaskedUninterruptible
-                         }
-            schedule thread' simstate
 
-          NotLocked -> do
-            -- Acquire lock
-            writeSTRef isLockedRef (Locked tid)
-            let -- Remove the timer from the queue
-                timers' = PSQ.delete tmid timers
-                -- Run the continuation
-                thread' = thread { threadControl = ThreadControl (k (Just x)) ctl' }
-            schedule thread' simstate { timers = timers' }
+        -- We cannot do `tryPutMVar` in the `treadAction`, because we need to
+        -- know if the `lock` is empty right now when we still have the frame.
+        v <- execTryPutTMVar lock_ undefined
+        let -- Kill the assassin throwing thread then unmask exceptions and
+            -- carry on the continuation
+            threadAction :: IOSim s ()
+            threadAction =
+              if v then unsafeUnregisterTimeout tmid
+                   else atomically (takeTMVar lock) >>= killThread
+
+            thread' =
+              thread { threadControl =
+                        ThreadControl (case threadAction of
+                                        IOSim k' -> k' (\() -> k (Just x)))
+                                      ctl'
```

Comment:
Why do you need pattern matching here?

Reply:
We need `SimA` not `IOSim`.

Reply:
What do you mean with that? What about the other places where we don't do pattern match? It is not clear to me why in this case we need to do this

Reply:
because in other palces we use `SimA` to write thee syntax tree.

Reply:
Oh nevermind forgot that threadAction existed and it's of type IOSim.